main/src/ca/uwaterloo/flix/MutationTester.scala

```diff
@@ -0,0 +1,398 @@
+    }
+    private def mutateRoot(root: TypedAst.Root, testee: String): List[(Symbol.DefnSym, List[TypedAst.Def])] = {
+        val defs = root.defs
+        val defSyms = root.modules.filter(pair => (pair._1.toString.equals(testee))).values.toList.flatten
+        mutateDefs(defs, defSyms).flatten
+    }
+
+
+    private def mutateDefs(defs: Map[Symbol.DefnSym, TypedAst.Def], defSyms: List[Symbol]) = {
+        defs.toList.map(d => (d._1, d._2) match {
+            case (s, fun) =>
+                if (defSyms.contains(s)) {
+                    val mutExps = mutateExpr(fun.exp)
+                    val mutDefs = mutExps.map(mexp => fun.copy(exp = mexp))
+                    Some(d._1 -> mutDefs)
+                } else None
+            case _ => None
+        })
+    }
+
+
+    private def mutationPermutations[A](exps: List[A], mutationFunc: (A => List[A])): List[List[A]] = {
+        def mutate(exps: List[A], index: Int) : List[A] = (exps, index) match {
+            case (x::Nil, 0) => mutationFunc(x)
+            case (x::xs, 0) => mutationFunc(x)
+            case (x::xs, n) => mutate(xs, n-1)
+            case (_, _) => Nil
+        }
+
+        def replace(index: Int, mutation: A, exps: List[A]): List[A] =
+            (index, exps) match {
+                case (0, _::Nil) => mutation :: Nil
+                case (0, x::xs) => mutation::xs
+                case (n, x::xs) => x :: replace(n-1, mutation, xs)
+                case (_,_) => Nil
+        }
+
+        var perms: List[List[A]] = Nil
+
+        for (i <- exps.indices) {
+            val mutations = mutate(exps, i)
+            for(m <- mutations){
+                perms = replace(i, m, exps) :: perms
+            }
+
+        }
+        perms
+
+    }
+    /**
+        // var doesn't contain a subtree, and we don't mutate them so we don't need to return them
+        case Expr.Var(_, _, _) => Nil
+      */
+    private def mutateExpr(e: TypedAst.Expr): List[TypedAst.Expr] = e match {
+        case Expr.Cst(cst, tpe, loc) =>
+            mutateCst(cst).map(m => Expr.Cst(m, tpe, loc))
+        case original@Expr.Var(_, _, _) => original :: Nil
+        case original@Expr.Def(sym, tpe, loc) => original :: Nil
+        case original@Expr.Sig(sym, tpe, loc) => original :: Nil
+        case original@Expr.Hole(sym, tpe, loc) => original :: Nil
+        case original@Expr.HoleWithExp(exp, tpe, eff, loc) => mutateExpr(exp).map(m => original.copy(exp = m))
+        case original@Expr.OpenAs(symUse, exp, tpe, loc) => mutateExpr(exp).map(m => original.copy(exp = m))
+        case original@Expr.Use(sym, alias, exp, loc) => mutateExpr(exp).map(m => original.copy(exp = m))
+        case original@Expr.Lambda(fparam, exp, _, _) =>
+            mutateExpr(exp).map(m => original.copy(exp = m))
+
+        case original@Expr.Apply(exp, exps, tpe, eff, loc) =>
+            mutationPermutations(exps, mutateExpr).map(mp => original.copy(exps = mp))
```

Comment:
Woops! What about the types? This is OK if the arguments have the same type, e.g. String and String. But what if the arguments have types String and Int??